codegen/src/next/mod.rs

```diff
@@ -1,13 +1,189 @@
 use proc_macro_error::emit_error;
 use quote::{quote, ToTokens};
-use syn::{Ident, LitStr};
+use syn::{
+    parenthesized, parse::Parse, parse2, punctuated::Punctuated, token::Paren, Expr, File, Ident,
+    Item, ItemFn, Lit, LitStr, Token,
+};
 
+#[derive(Debug, Eq, PartialEq)]
 struct Endpoint {
     route: LitStr,
     method: Ident,
     function: Ident,
 }
 
+#[derive(Debug, Eq, PartialEq)]
+struct Parameter {
+    key: Ident,
+    equals: Token![=],
+    value: Expr,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+struct Params {
+    params: Punctuated<Parameter, Token![,]>,
+    paren_token: Paren,
+}
+
+impl Parse for Parameter {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        Ok(Self {
+            key: input.parse()?,
+            equals: input.parse()?,
+            value: input.parse()?,
+        })
+    }
+}
+
+impl Parse for Params {
+    fn parse(input: syn::parse::ParseStream) -> syn::Result<Self> {
+        let content;
+        Ok(Self {
+            paren_token: parenthesized!(content in input),
+            params: content.parse_terminated(Parameter::parse)?,
+        })
+    }
+}
+
+impl Endpoint {
+    fn from_item_fn(item: &mut ItemFn) -> Option<Self> {
+        let function = item.sig.ident.clone();
+
+        let mut endpoint_index = None;
+
+        // Find the index of an attribute that is an endpoint
+        for index in 0..item.attrs.len() {
+            // The endpoint ident should be the last segment in the path
+            if let Some(segment) = item.attrs[index].path.segments.last() {
+                if segment.ident.to_string().as_str() == "endpoint" {
+                    if endpoint_index.is_some() {
+                        emit_error!(
+                            item,
+                            "extra endpoint attribute";
+                            hint = "There should only be one endpoint annotation per handler function."
```

Comment:
Not sure why the limitation? The same handler could be applied to multiple routes or methods.

Though TBF, since this PR only has the proc macro part of this, we can wait until we know how we use it in building routers to make a change to this.

Reply:
Regarding the `::app!`, I believe Pieter said we would use a build.rs for this down the line, but I'm not familiar with the details yet. 

I know that having several handlers with different methods for the same route is common, but is using the same handler for different routes common? Common or not though, it's certainly something we could handle, maybe by returning a Vec<Endpoint> from this and then validating that there aren't any invalid combinations. Doing this when we know more about the final setup sounds like a good plan :+1: